src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs

```diff
@@ -2306,6 +2311,886 @@ internal unsafe void CM(B b)
+{
+    static void M1<T>(T t) => Console.Write(t);
+    static void Main()
+    {
+        delegate*<int, void> ptr = &C.M1<int>;
+        ptr(1);
+        ptr = &C.M1;
+        ptr(2);
+    }
+}", expectedOutput: "12");
+
+            verifier.VerifyIL("C.Main()", expectedIL: @"
+{
+  // Code size       29 (0x1d)
+  .maxstack  2
+  .locals init (delegate*<int,void> V_0)
+  IL_0000:  ldftn      ""void C.M1<int>(int)""
+  IL_0006:  stloc.0
+  IL_0007:  ldc.i4.1
+  IL_0008:  ldloc.0
+  IL_0009:  calli      0x3
+  IL_000e:  ldftn      ""void C.M1<int>(int)""
+  IL_0014:  stloc.0
+  IL_0015:  ldc.i4.2
+  IL_0016:  ldloc.0
+  IL_0017:  calli      0x3
+  IL_001c:  ret
+}
+");
+        }
+
+        [Fact]
+        public void ReducedExtensionMethod()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe static class CHelper
+{
+    public static void M1(this C c) {}
+}
+unsafe class C
+{
+    static void M(C c)
+    {
+        delegate*<C, void> ptr = &c.M1;
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (10,34): error CS8757: No overload for 'M1' matches function pointer 'delegate*<C,void>'
+                //         delegate*<C, void> ptr = &c.M1;
+                Diagnostic(ErrorCode.ERR_MethFuncPtrMismatch, "&c.M1").WithArguments("M1", "delegate*<C,void>").WithLocation(10, 34)
+            );
+        }
+
+        [Fact]
+        public void UnreducedExtensionMethod()
+        {
+            var verifier = CompileAndVerifyFunctionPointers(@"
+#pragma warning suppress CS0414 // Field never used
+using System;
+unsafe static class CHelper
+{
+    public static void M1(this C c) => Console.Write(c.i);
+}
+unsafe class C
+{
+    public int i;
+    static void Main()
+    {
+        delegate*<C, void> ptr = &CHelper.M1;
+        var c = new C();
+        c.i = 1;
+        ptr(c);
+    }
+}", expectedOutput: "1");
+
+            verifier.VerifyIL("C.Main()", expectedIL: @"
+{
+  // Code size       28 (0x1c)
+  .maxstack  3
+  .locals init (C V_0, //c
+                delegate*<C,void> V_1)
+  IL_0000:  ldftn      ""void CHelper.M1(C)""
+  IL_0006:  newobj     ""C..ctor()""
+  IL_000b:  stloc.0
+  IL_000c:  ldloc.0
+  IL_000d:  ldc.i4.1
+  IL_000e:  stfld      ""int C.i""
+  IL_0013:  stloc.1
+  IL_0014:  ldloc.0
+  IL_0015:  ldloc.1
+  IL_0016:  calli      0x5
+  IL_001b:  ret
+}
+");
+        }
+
+        [Fact]
+        public void BadScenariosDontCrash()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    static void M1() {}
+    static void M2()
+    {
+        &delegate*<void> ptr = &M1;
+    }
+    static void M3(C c)
+    {
+        // Wrong accessiblity, can't reference reduced extension method
+        delegate*<C, void> ptr = c.H1;
+    }
+}
+static class CHelper
+{
+    // Wrong accessiblity
+    public static void H1(this C c) {}
+}");
+
+            comp.VerifyDiagnostics(
+                // (7,18): error CS1514: { expected
```

Comment:
>// (7,18): error CS1514: { expected [](start = 16, length = 35)

Consider separating the parse error with the extension method test, if that (the extension method scenario) tests something new. Otherwise, consider removing it completely.